content/_code-samples/trade-in-the-decentralized-exchange/js/trade-in-the-dex.js

```diff
@@ -0,0 +1,227 @@
+// Stand-alone code sample for "trade in the decentralized exchange" tutorial:
+// https://xrpl.org/trade-in-the-decentralized-exchange.html
+
+// Dependencies for Node.js.
+// In browsers, use <script> tags as in the example demo.html.
+if (typeof module !== "undefined") {
+  // Use var here because const/let are block-scoped to the if statement.
+  var xrpl = require('xrpl')
+  var BigNumber = require('bignumber.js')
+}
+
+// Connect ---------------------------------------------------------------------
+async function main() {
+  const client = new xrpl.Client('wss://s.altnet.rippletest.net:51233')
+  console.log("Connecting to Testnet...")
+  await client.connect()
+
+  // Get credentials from the Testnet Faucet -----------------------------------
+  console.log("Requesting address from the Testnet faucet...")
+  const wallet = (await client.fundWallet()).wallet
+  console.log(`Got address ${wallet.address}.`)
+  // To use existing credentials, you can load them from a seed value, for
+  // example using an environment variable as follows:
+  // const wallet = xrpl.Wallet.fromSeed(process.env['MY_SEED'])
+
+  // Define the proposed trade. ------------------------------------------------
+  // Technically you don't need to specify the amounts (in the "value" field)
+  // to look up order books using book_offers, but for this tutorial we reuse
+  // these variables to construct the actual Offer later.
+  const we_want = {
+    currency: "TST",
+    issuer: "rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd",
+    value: "25"
+  }
+  const we_spend = {
+    currency: "XRP",
+    value: xrpl.xrpToDrops(25*10*1.15) // 25 TST * 10 XRP per TST * 15% fx cost
+  }
+  // "Quality" is defined as TakerPays ÷ TakerGets. The lower the "quality"
+  // number, the better the proposed exchange rate is for the taker.
+  // The quality is rounded to a number of significant digits based on the
+  // issuer's TickSize value (or the lesser of the two for token-token trades.)
+  const proposed_quality = BigNumber(we_spend.value) / BigNumber(we_want.value)
```

Comment:
This code doesn't actually round off the quality based on the TickSize. That means it'll give inaccurate results if the token in question uses a small enough TickSize. I actually ran into this problem initially, but worked around it by resetting rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd's TickSize setting to the default.

I don't want to make the tutorial more complicated by adding a bunch of stuff to deal with this, but it would be nice to know if/how you could use BigNumber to round the quality's significant digits (note: _not_ to a specific scale) to account for TickSize settings.